src/binary/ibc-setup/commands/channels.ts

```diff
@@ -0,0 +1,147 @@
+  readonly interactive: boolean;
+};
+
+export type Options = {
+  readonly home: string;
+  readonly chain: string;
+  readonly mnemonic: string | null;
+  readonly port: string | null;
+};
+
+export async function channels(flags: Flags) {
+  const home = resolveHomeOption({ homeFlag: flags.home });
+  const app = loadAndValidateApp(home);
+  const keyFile = resolveKeyFileOption({ keyFileFlag: flags.keyFile, app });
+  const chain = resolveRequiredOption('chain')(
+    flags.chain,
+    process.env.RELAYER_CHAIN
+  );
+  const port = resolveOption(flags.port, process.env.RELAYER_PORT);
+
+  const mnemonic = await resolveMnemonicOption(
+    {
+      interactiveFlag: flags.interactive,
+      mnemonicFlag: flags.mnemonic,
+      keyFile,
+      app,
+    },
+    true // mnemonic is optional
+  );
+
+  const options: Options = {
+    home,
+    chain,
+    mnemonic,
+    port,
+  };
+
+  await run(options);
+}
+
+function assureMnemonic(mnemonic: string | null) {
+  if (!mnemonic) {
+    return generateMnemonic();
+  }
+
+  return mnemonic;
+}
+
+function stateAsText(state: State) {
+  switch (state) {
+    case State.STATE_CLOSED:
+      return 'Closed';
+
+    case State.STATE_INIT:
+      return 'Init';
+
+    case State.STATE_OPEN:
+      return 'Open';
+
+    case State.STATE_TRYOPEN:
+      return 'Tryopen';
+
+    case State.STATE_UNINITIALIZED_UNSPECIFIED:
+      return 'UninitializedUnspecified';
+
+    case State.UNRECOGNIZED:
+    default:
+      return 'Unrecognized';
+  }
+}
+
+function appendSpaces(value: string) {
+  const spaces = new Array(24 - value.length).fill(' ').join('');
+  return `${value}${spaces}`;
+}
+
+export async function run(options: Options) {
+  const registryFilePath = path.join(options.home, registryFile);
+  const registry = loadAndValidateRegistry(registryFilePath);
+
+  const chain = registry.chains[options.chain];
+  if (!chain) {
+    throw new Error(`Chain ${options.chain} not found in ${registryFilePath}.`);
+  }
+
+  const mnemonic = assureMnemonic(options.mnemonic);
+
+  const client = await createClient(mnemonic, {
+    prefix: chain.prefix,
+    rpc: chain.rpc,
+    hd_path: chain.hd_path,
+  });
+
+  const {
+    channels: allChannels,
+  } = await client.query.ibc.channel.allChannels();
+
+  const channels = allChannels
+    .filter(
+      (channel) => (options.port ? channel.portId === options.port : true) // don't filter if port is not specified
+    )
+    .map((channel) =>
+      [channel.channelId, channel.portId, stateAsText(channel.state)]
+        .map(appendSpaces)
+        .join('')
+    );
+
+  if (!channels.length) {
+    const conditionalPortInfo = options.port
+      ? ` on port "${options.port}".`
+      : '.';
+    console.log(
+      `Found no channels for chain "${options.chain}"${conditionalPortInfo}`
+    );
+
+    return;
+  }
+
+  const output = [
+    ['CHANNEL_ID', 'PORT', 'STATE'].map(appendSpaces).join(''),
+    ...channels,
+  ].join(os.EOL);
+
+  console.log(output);
```

Comment:
@ethanfrey 

The spec says:
>It shows all channels, both open, as well as channels in the handshaking process or closed ones (with a comment on non-open ones)

but here we turn every state into text, like:
<img width="477" alt="image" src="https://user-images.githubusercontent.com/79154001/110549709-c77ae500-8132-11eb-9479-faee5b3dfb80.png">

Is it okay?

Reply:
That output looks quite nice